### ISSUE TITLE
Monitor atlas2 tool

### ISSUE THREAD

**Issue by cdaniel** [cdaniel/atlas2#175]
The tool goes down from time to time, looks like bluemix issue, since the CF app is up and running, but route is not served properly.

# Requirements:
The solution should:
* make end to end scripts creation as easy as possible, ideally through recording user actions in the browser. However, tests creation must not rely on tools/libraries that are not free/open for everyone.
* let the developer to execute tests locally, against his local tool instance
* let me to use it as availability monitoring (when executed on some server, service plan might be necessary for this)

# Findings:
## 1. Selenium IDE
Supports selenium scripts in a form of HTML or zip file. But:
* manually encoding tests in HTML files seems to be very wrong
* Selenium IDE is a dead duck

## 2. Ghost Inspector
Seems to be able to record actions and export them in the selenium format. IBM Bluemix Monitoring still could be used (given it would be possible to customise input data). Work would require:

- [x] ~~verify customisations and running scripts locally~~
- [x] ~~find out how to run selenium tests in travis~~

Turns out IBM Bluemix cannot reliably runs those scripts. Hence, the two first approaches are excluded.

## 3. Nightwatch
De facto js standard of UI testing. Requires extensive configuration (may run locally) + running tests continously is expensive ($149 per month), therefore cannot be used for Atlas2 monitoring. Plus, the only editor is snaptest, it has limited functionality, and tests do not go well with existing mocha tests. It would be necessary to:
-  integrate nightwatch and mocha
-  write tests locally (in JS instead of recording)
-  integrate travis with those tests
-  have simple monitoring in Bluemix in other technique.

**Comment by jhngrant** [comment]
@cdaniel Point 1 mentions coding tests in HTML is wrong. Agreed. You are aware Selenium can be scripted in most languages? I'm running this past you because I think it's hard to beat. These days vendor WebDrivers - such as Marionette for Firefox and the Chrome WebDriver (https://sites.google.com/a/chromium.org/chromedriver/downloads) are robust and well maintained.

With such a wide range of devices out there, some touch enabled, a browser session recorder is a mighty challenge. Even though it would be really handy and a timesaver for end to end testing of SPAs, I doubt such a tool is now viable to reach production quality.

IMHO the WebDriver protocol is the de facto standard for remote UI testing. There's a W3C candidate recommendation - https://www.w3.org/TR/webdriver/

Just in case you aren't aware WebdriverIO - http://webdriver.io/ - may offer additional functionality over the native Selenium Node driver.

**Comment by cdaniel** [comment]
@jhngrant, many thanks for your comment. It makes a couple of things clear:

- I have to abandon Bluemix Availability Monitoring - it accepts Selenium scripts only in the form of html. Unfortunately, I cannot find a good replacement for it. 
- The choice of the framework boils down to [WebDriverIO and Nightwatch](https://stackoverflow.com/questions/35981605/what-is-the-difference-between-nightwatch-js-and-webdriver-io). Indeed, I was not aware of the webdriver.io (which seems to have integration with ).
- I have to ditch the idea of using the tool. I tried to use snaptest.io to record actions, it worked, and would be a big argument in favour of Nightwatch if it was more mature.

In addition to all of that, webdriver.io seems to have a good support for local testing, cloud (Testing Bot) and Travis. Thanks for bringing that!

**Comment by jpospychala** [comment]
I would try  this new thing from google: https://github.com/GoogleChrome/puppeteer

**Comment by cdaniel** [comment]
@jpospychala thanks for puppeteer. It seems to be something fresh, although it has no associated ecosystem, and is limited to chrome. But I will try to use it to replace PhantomJs which I use to generate maps for download.

**Comment by cdaniel** [comment]
Nightwatch seems to be a dying duck. The main developer has not been very active for about a year.

**Comment by cdaniel** [comment]
I think it is time to close this one (with the pull request #175). Short summary:
* WebdriverIO is used, configured for chrome and firefox.
* Selenium-standalone manages selenium packages. Java 1.8 must be available, though.
* I failed to integrate wdio with mocha. `wdio` (the standalone command) does a lot of work in the background, and, unforunately, requires different approach to writing tests than mocha does. Because of that, mocha runs unit tests, and `wdio` e2e tests.
* with the help of xvbf, I was able to run all tests on Travis. Maybe, in the future, I will use Sauce Labs for more browsers.
* with the help of Travis cron jobs, I was able to set up repeated e2e testing (aka monitoring) of the atlas2.wardleymaps.com. It is not perfect, tests are executed rarely (every day), but will do for now,

@jpospychala, @jhngrant thank you again for your input!
